### PR TITLE
[GPU] Optimize sdpa_opt_finalization_stage

### PR DESCRIPTION
**Pull request by zaixing-wang**
### Details:
 - to update Finalize stage GWS format from (x, x, sub_group_size) to (x, x, head_size), and update LWS format from (1, 1, sub_group_size) to (1, 1, head_size). 
 
- to replace exp_sum and max_logit with max_logits_u_exp_sum_slm. 
- to put the division by global_max at the end.

### Performance:
- qwen2-7B’s throughput increased from 40.6 qps to 42.18 qps in Arc770.

### Tickets:
 - CVS-159469

### CODE REVIEW

**Comment by ceciliapeng2011** [comment]
> to update Finalize stage GWS format from (x, x, sub_group_size) to (x, x, head_size), and update LWS format from (1, 1, sub_group_size) to (1, 1, head_size).
to replace exp_sum and max_logit with max_logits_u_exp_sum_slm.
to put the division by global_max at the end.

Would you please comment why do you have these updates?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl [hunk 1]
```diff
@@ -1596,185 +1596,170 @@ KERNEL(sdpa_opt_finalization_stage)(
...
+        if (sglid == 0) {
+            tmp_slm[sgid] = local_max_logit;
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
 
-        SOFTMAX_ACCUMULATOR_TYPE global_max = sub_group_reduce_max(local_max_logit);
+        if (sglid < HEAD_SIZE / SUBGROUP_SIZE) {
+            local_max_logit = tmp_slm[sglid];
+        }
+        local_max_logit = sub_group_reduce_max(local_max_logit);
```
Comment (ceciliapeng2011) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl:1628]: It is not "local" any more..it is "global". If there is no register spilling, better keep its name "global_max".

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl [hunk 2]
```diff
@@ -1596,185 +1596,170 @@ KERNEL(sdpa_opt_finalization_stage)(
...
+            local_max_logit = SOFTMAX_ACCUMULATOR_MAX_FUNC(local_max_logit, max_logits_u_exp_sum_slm[i]);
         }
+        local_max_logit = sub_group_reduce_max(local_max_logit);
+        if (sglid == 0) {
+            tmp_slm[sgid] = local_max_logit;
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
 
-        SOFTMAX_ACCUMULATOR_TYPE global_max = sub_group_reduce_max(local_max_logit);
+        if (sglid < HEAD_SIZE / SUBGROUP_SIZE) {
```
Comment (ceciliapeng2011) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl:1625]: What if HEAD_SIZE/SUBGROUP_SIZE > SUBGROUP_SIZE? Is it possible?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl [hunk 3]
```diff
@@ -1596,185 +1596,170 @@ KERNEL(sdpa_opt_finalization_stage)(
...
+            local_exp_sum = tmp_slm[sglid];
         }
 
-        SOFTMAX_ACCUMULATOR_TYPE global_sum = sub_group_reduce_add(local_exp_sum);
-
-        for (uint head_size_idx = 0; head_size_idx < HEAD_SIZE / SUBGROUP_SIZE; head_size_idx++) {
-            SOFTMAX_ACCUMULATOR_TYPE acc = 0.0f;
-            for (uint partition_idx = 0; partition_idx < num_of_partitions; partition_idx++) {
+        local_exp_sum = sub_group_reduce_add(local_exp_sum);
```
Comment (ceciliapeng2011) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl:1647]: similar

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl [hunk 4]
```diff
@@ -1596,185 +1596,170 @@ KERNEL(sdpa_opt_finalization_stage)(
...
-                    TO_SOFTMAX_ACCUMULATOR_TYPE(sub_group_broadcast(exp_sum[partition_idx / SUBGROUP_SIZE], partition_idx % SUBGROUP_SIZE)) /
-                    TO_SOFTMAX_ACCUMULATOR_TYPE(global_sum);
-            }
-            const uint out_offset = b0_idx * (NUM_HEADS * TARGET_SEQ_LEN * HEAD_SIZE) +
-                                    b1_idx * (TARGET_SEQ_LEN * HEAD_SIZE) +
-                                    target_seq_idx * (HEAD_SIZE) +
-                                    (head_size_idx * SUBGROUP_SIZE + sglid);
-
-            output[out_offset] = TO_OUTPUT_TYPE(acc);
+                acc += TO_SOFTMAX_ACCUMULATOR_TYPE(tmp_out[tmp_out_offset]) * TO_SOFTMAX_ACCUMULATOR_TYPE(max_logits_u_exp_sum_slm[partition_idx]);
```
Comment (ceciliapeng2011) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl:1655]: Do you forget out_val here?

**Comment by ceciliapeng2011** [comment]
@sshlyapn two questions to the original implementation of sdpa second token -
1. why do we have branches like "if (num_of_partitions <= SUBGROUP_SIZE * REG_VERSION_MAX_VALUES_PER_WI_LOWER) {" and "if (num_of_partitions <= SUBGROUP_SIZE * REG_VERSION_MAX_VALUES_PER_WI) {"? Any good for performance?
2. The output_offset is calculated assuming a layout like BHLS, without considering transpose. Isn't any good to fuse the transpose node after  sdpa?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl [hunk 5]
```diff
@@ -1596,185 +1596,170 @@ KERNEL(sdpa_opt_finalization_stage)(
...
-                                         target_seq_idx * (num_of_partitions) +
-                                         partition_idx;
-            const uint max_logit_offset = exp_sums_offset;
-
-            if (partition_idx < num_of_partitions) {
-                exp_sum[i] = exp_sums[exp_sums_offset];
-                max_logit[i] = max_logits[max_logit_offset];
-                local_max_logit = SOFTMAX_ACCUMULATOR_MAX_FUNC(local_max_logit, max_logit[i]);
-            }
+        SOFTMAX_ACCUMULATOR_TYPE max_logits_u_exp_sum_slm[REG_VERSION_MAX_VALUES_PER_WI_LOWER] = {SOFTMAX_ACCUMULATOR_VAL_MIN};
```
[the remark this answers is not in the document]
Comment (zaixing-wang) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl]: Done. I use slm now.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl (hunk 1 above)
Comment (zaixing-wang) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl]: done

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl (hunk 2 above)
Comment (zaixing-wang) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl:1626]: Done, add a condition to solve this case.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl (hunk 3 above)
Comment (zaixing-wang) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl]: done

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl (hunk 4 above)
Comment (zaixing-wang) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl]: done

**Comment by sshlyapn** [comment]
> @sshlyapn two questions to the original implementation of sdpa second token -
> 
> 1. why do we have branches like "if (num_of_partitions <= SUBGROUP_SIZE * REG_VERSION_MAX_VALUES_PER_WI_LOWER) {" and "if (num_of_partitions <= SUBGROUP_SIZE * REG_VERSION_MAX_VALUES_PER_WI) {"? Any good for performance?
> 2. The output_offset is calculated assuming a layout like BHLS, without considering transpose. Isn't any good to fuse the transpose node after  sdpa?

@ceciliapeng2011
1. There were register spills for MTL with the REG_VERSION_MAX_VALUES_PER_WI variable usage, while it worked fine for other platforms. So, I added a lighter version which also covers most of the cases but doesn't suffer from register spills.
2. Yeah, the current SDPA primitive has everything set up for output transpose fusion, but output fusion hasn't been implemented at the kernel level yet. Despite the fact that the total time portion of such transposes is not that big, we can probably still see some improvements in particular cases. We just haven't had enough time to implement this

**Review by ceciliapeng2011 — APPROVED** [review]
LGTM, looking forward to model accuracy and perf report.

**Review by sshlyapn — APPROVED** [review]
LGTM

**Review by vladimir-paramuzov — APPROVED** [review]
